For this exercise
- Write a function (guessing_game) that takes no arguments.
- When run, the function chooses a random integer between 0 and 100 (inclusive).
- Then ask the user to guess what number has been chosen.
- Each time the user enters a guess, the program indicates one of the following:
  - Too high
  - Too low
  - Just right
- If the user guesses correctly, the program exits. Otherwise, the user is asked to
try again.
- The program only exits after the user guesses correctly

In [ ]:
from random import randint

def guessing_game():
    number = randint(0, 100)
    while True:
      guess = int(input('Insert your guess: '))
      if guess == number:
        print('Just right')
        break
      elif guess < number:
        print('Too low')
      else:
        print('Too high')


Modify this program, such that it gives the user only three chances to guess the
correct number. If they try three times without success, the program tells them
that they didn’t guess in time and then exits.

In [ ]:
def guessing_game():
    number = randint(0, 100)
    for trial in range(3):
        guess = int(input('Insert your guess: '))
        if guess == number:
          print('Just right')
          break
        elif guess < number:
          print('Too low')
        else:
          print('Too high')
guessing_game()

Not only should you choose a random number, but you should also choose a random number base, from 2 to 16, in which the user should submit their input. If the user inputs “10” as their guess, you’ll need to interpret it in the correct number base; “10” might mean 10 (decimal), or 2 (binary), or 16
(hexadecimal).

Try the same thing, but have the program choose a random word from the dictionary, and then ask the user to guess the word. (You might want to limit yourself to words containing two to five letters, to avoid making it too horribly
difficult.) Instead of telling the user that they should guess a smaller or larger number, have them choose an earlier or later word in the dict

EXERCISE 2 Summing numbers

The splat operator, designing functions that can accept a variable number of arguments

In [ ]:
def mysum(*numbers):
    sum = 0
    for number in numbers:
        sum = sum + number
    return sum

mysum(4, 6, 5, 5, 8, 9)

In [ ]:
def average(*numbers):
    print(numbers)
    count = len(numbers)
    sum = 0
    for number in numbers:
        sum = sum + number
    return sum / count

average(4, 1, 2, 5, 6)

The built-in version of sum takes an optional second argument, which is used as the starting point for the summing. (That’s why it takes a list of numbers as its first argument, unlike our mysum implementation.) So sum([1,2,3], 4) returns 10, because 1+2+3 is 6, which would be added to the starting value of 4. Reimplement your mysum function such that it works in this way. If a second argument is not provided, then it should default to 0. Note that while you can write a function in Python 3 that defines a parameter after *args, I’d suggest avoiding it and just taking two arguments—a list and an optional starting point.

In [ ]:
def sum(iterable, seed: int = 0):
    summation = seed
    for number in iterable:
        summation = summation + number
    return summation
sum((1, 2, 4), 10)

Write a function that takes a list of numbers. It should return the average (i.e., arithmetic mean) of those numbers.

In [ ]:
from typing import List
def avg(numbers: List[int]):
    num_items = len(numbers)
    total = sum(numbers)
    return total / num_items
avg([4])

Write a function that takes a list of words (strings). It should return a tuple containing three integers, representing the length of the shortest word, the length of the longest word, and the average word length.

In [ ]:
def shortest(names):
    short = len(names[0])
    for current_name in names[1:]:
        if len(current_name) < short:
            short = len(current_name)
    return short

def longest(names):
    long = len(names[0])
    for current_name in names[1:]:
        if len(current_name) > long:
            long = len(current_name)
    return long


def avg_word(names):
    num_items = len(names)
    sum = 0
    for name in names:
        sum = sum + len(name)
    return sum / num_items

def solution(names):
    return (shortest(names), longest(names), avg_word(names))

solution(['kasozi', 'vincent', 'isaac'])

Write a function that takes a list of Python objects. Sum the objects that either are integers or can be turned into integers, ignoring the others

In [ ]:
def sum_objs(objects):
    '''This is my function.
       This is the second line
    '''
    sum = 0
    for obj in objects:
        try:
         if isinstance(obj, int):
            sum = sum + obj
         else:
            sum = sum + int(obj)
        except:
           continue
    return sum

sum_objs([True, False, '5', 'Vanessa', 'Kasozi', 8])

Write a function (run_timing) that asks how long it took for you to run 10 km. The function continues to ask how long (in minutes) it took for additional runs, until the user presses Enter. At that point, the function exits—but only after calculating and displaying the average time that the 10 km runs took.

In [ ]:
def run_timing():
    days = 0
    sum = 0
    while True:
        response = input('Insert the time for 10km in seconds: ')
        if response == '':
            break
        sum = sum + float(response)
        days = days + 1        
    return sum / days    
run_timing()

Write a function that takes a float and two integers (before and after). The function should return a float consisting of before digits before the decimal point and after digits after. Thus, if we call the function with 1234.5678, 2 and
3, the return value should be 34.567.

In [ ]:
def format(number: float, before, after):
    string = str(number)
    left, right = string.split('.')
    new_left = left[-before:]
    new_right = right[:after]
    return float(f'{new_left}.{new_right}')

format(1234.5678, 1, 3)

In [ ]:
name = 'Vanessa'
# get one item from the string
print(name[4])

# get the last letter in the string
print(name[-1])

#get the first 2 letters
print(name[:2])

#get the first 5 letters
print(name[:5])

#get the last two letters
print(name[-2:])

# get the last 5 letters
print(name[-5:])

s = '1234'
print(s[-2:])

t = '5678'
print(t[0:3])


Explore the Decimal class (http://mng.bz/oPVr), which has an alternative floating-point representation that’s as accurate as any decimal number can be. Write a function that takes two strings from the user, turns them into decimal
instances, and then prints the floating-point sum of the user’s two inputs. In other words, make it possible for the user to enter 0.1 and 0.2, and for us to get 0.3 back.

In [ ]:
from decimal import Decimal

def solution():
    first = Decimal(input('Insert the first string: '))
    second = Decimal(input('Insert the second string: '))
    return float(first + second)
solution()

In [ ]:
def convert10(input, base):
  decnum = 0
  for power, digit in enumerate(reversed(input)): 
     decnum += int(digit, base) * (base ** power) 
  return decnum

Reimplement the solution for this exercise such that it doesn’t use the int function at all, but rather uses the built-in ord and chr functions to identify the character. This implementation should be more robust, ignoring characters that aren’t legal for the entered number base.

Write a program that asks the user for their name and then produces a “name triangle”: the first letter of their name, then the first two letters, then the first three, and so forth, until the entire name is written on the final line. 

In [ ]:
def name_triangle(name):
    for (idx, _) in enumerate(name):
        print(name[0: idx + 1])
name_triangle('Vanessa')

Pig Latin (http://mng.bz/YrON) is a common children’s “secret” language in English speaking countries. (It’s normally secret among children who forget that their parents
were once children themselves.) The rules for translating words from English into Pig
Latin are quite simple:
- If the word begins with a vowel (a, e, i, o, or u), add “way” to the end of the
word. So “air” becomes “airway” and “eat” becomes "eatway."
- If the word begins with any other letter, then we take the first letter, put it on
the end of the word, and then add "ay." Thus, “python” becomes “ythonpay”
and “computer” becomes “omputercay.”


In [2]:
def is_vowel(letter):
    return letter in 'aeiouAIEOU'

def pig_latin(word: str) -> str:
    if is_vowel(word[0]):
        return f'{word}way'
    else:
        return f'{word[1:]}{word[0]}ay'

def pig_latin_pun(word: str) -> str:
    if is_vowel(word[0]):
        return f'{word[:-1]}way{word[-1]}'
    else:
        return f'{word[1:-1]}{word[0]}ay{word[-1]}'
    


Handle capitalized words—If a word is capitalized (i.e., the first letter is capitalized, but the rest of the word isn’t), then the Pig Latin translation should be similarly capitalized.

In [ ]:
def pig_latin_caps(word: str) -> str:
    if word[0].islower():
        return pig_latin(word)
    else:
        return pig_latin(word).capitalize()
pig_latin_caps('air')

Handle punctuation—If a word ends with punctuation, then that punctuation should be shifted to the end of the translated word.

In [ ]:
import string
def pig_latin_punc(word: str) -> str:
    if word[-1] in string.punctuation:
        return pig_latin_pun(word)
    else:
        return pig_latin_caps(word)
pig_latin_punc('air')

Now that you’ve successfully written a translator for a single English word, let’s make things more difficult: translate a series of English words into Pig Latin. Write a function called pl_sentence that takes a string containing several words, separated by spaces. (To make things easier, we won’t actually ask for a real sentence. More specifically, there will be no capital letters or punctuation.)

In [ ]:
def pl_sentence(sentence: str) -> str:
    foo = []
    for word in sentence.split():
        foo.append(pig_latin(word))
    return ' '.join(foo)

pl_sentence('this is a test translation')

Take a text file, creating (and printing) a nonsensical sentence from the nth word on each of the first 10 lines, where n is the line number.

In [ ]:
city = 'New York'
city

In [ ]:
city = "New York"
city

In [ ]:
number = str(8)
number

String Concatenation

In [ ]:
first_name = 'Mirembe'
second_name = 'Vanessa'
full_name = f'{first_name} {second_name}'
full_name

Selecting a char

In [ ]:
name = 'Mirembe Vanessa'
name[0]

Size of a String

In [ ]:
name = 'Mirembe Vanessa'
len(name)

Replacing a substring in a string

In [ ]:
name = 'Mirembe Vincent'
name.replace('Mirembe', 'Kasozi')

Counting substring occurences in a string

In [ ]:
'Vincent Vanessa'.count('Vin')

Repeating a String

In [ ]:
'Vincent' * 4

How to Split a String

In [ ]:
names = ['kasozi Vincent Vanessa', 'Mirembe Vanessa']
for name in names:
    first, middle = name.split(maxsplit=1)
    print(first, middle)

In [ ]:
'mIREMEBE vANESSA'.swapcase()

- input
- print
- int
- bool
- float
- str
- enumerate
- reversed
- sum
- help
- type
- len
- hex
- dir

Ubbi Dubbi

In [ ]:
def ubbi_dubbi(word: str) -> str:
    response = []
    for letter in word:
        if is_vowel(letter):
            response.append(f'ub{letter}')
        else:
            response.append(letter)
    return ''.join(response)

ubbi_dubbi('eat')

Handle capitalized words—If a word is capitalized (i.e., the first letter is capitalized, but the rest of the word isn’t), then the Ubbi Dubbi translation should be similarly capitalized.

In [ ]:
def ubbi_dubbi_caps(word: str) -> str:
    if word[0].isupper():
        return ubbi_dubbi(word).capitalize()
    else:
        return ubbi_dubbi(word)
ubbi_dubbi_caps('Eat')

Remove author names—In academia, it’s common to remove the authors’ names from a paper submitted for peer review. Given a string containing an article and a separate list of strings containing authors’ names, replace all names in the
article with _ characters.

In [ ]:
from typing import List
def replace_names(article: str, names: List[str]) -> str:
    for name in names:
        article = article.replace(name, '_' * len(name))
    return article

article = '''
     Makerere University .....
     Kasozi Vincent Mirembe Vanessa
      abstract:
      dskhlakfhkafh;ahf;sjkdfh;adsfjh;f
      sdfjakfjadflkjafljksdfakfkdfbakdfbak
'''


names = ['Kasozi Vincent', 'Mirembe Vanessa']

print(replace_names(article, names))

URL-encode characters—In URLs, we often replace special and nonprintable characters with a % followed by the character’s ASCII value in hexadecimal. For example, if a URL is to include a space character (ASCII 32, aka 0x20), we
replace it with %20. Given a string, URL-encode any character that isn’t a letter or number. For the purposes of this exercise, we’ll assume that all characters are indeed in ASCII (i.e., one byte long), and not multibyte UTF-8 characters. It might help to know about the ord (http://mng.bz/EdnJ) and hex (http://mng
.bz/nPxg) functions.

In [ ]:
import string
def url_encode(url):
    response = []
    for letter in url:
        if (letter in string.punctuation) or letter.isspace():
            s = hex(ord(letter)).replace('0x', '%')
            response.append(s)
        else:
            response.append(letter)
    return ''.join(response)

url_encode('hello world ')

 Sorting a string

In [ ]:
def sort_string(word: str) -> str:
    return ''.join(sorted(word))

def sort_capitalized(word: str) -> str:
    if word[0].isupper():
        return ''.join(sorted(word.lower())).capitalize()
    else:
        return sort_string(word)

sort_capitalized('vanessa')

Given the string “Tom Dick Harry,” break it into individual words, and then sort those words alphabetically. Once they’re sorted, print them with commas (,) between the names.

In [ ]:
def sort_sentence(sentence: str) -> str:
    response = []
    for word in sentence.split():
        response.append(sort_capitalized(word))
    return ', '.join(response)

sort_sentence('Tom Dick Harry')

Which is the last word, alphabetically, in a text file?

In [ ]:
def get_last(file: str) -> str:
    words = file.split()
    return sorted(words)[-1]

Which is the longest word in a text file?

In [ ]:
names = ['Vanessa', 'Alex','Kasozi']
sorted(names, key=len)[-1]

Write a function, firstlast, that takes a sequence (string, list, or tuple) and returns the first and last elements of that sequence, in a two-element sequence of the same type. So firstlast('abc') will return the string ac, while firstlast([1,2,3,4]) will return the list [1,4].

In [60]:
def first_last(seq):
    return seq[0:1] + seq[-1:]

first_last(tuple('abcde'))

('a', 'e')

Don’t write one function that squares integers, and another that squares floats.
Write one function that handles all numbers.

In [64]:
def square(number):
    return number ** 2
square(9.0)

81.0

Don’t write one function that finds the largest element of a string, another that does the same for a list, and a third that does the same for a tuple. Write just one function that works on all of them.

In [69]:
def maximum(seq):
    biggest = seq[0]
    for element in seq[1:]:
        if element > biggest:
            biggest = element
    return biggest

'z'

Write a function that takes a list or tuple of numbers. Return a two-element list, containing (respectively) the sum of the even-indexed numbers and the sum of the odd-indexed numbers. So calling the function as even_odd_sums([10, 20,
30, 40, 50, 60]), you’ll get back [90, 120].

In [78]:
def even_odds_sums(seq):
    evens = sum(seq[::2])
    odds = sum(seq[1::2])
    return [evens, odds]

even_odds_sums(range(10, 70, 10))

[90, 120]

Write a function that takes a list or tuple of numbers. Return the result of alternately adding and subtracting numbers from each other. So calling the function as plus_minus([10, 20, 30, 40, 50, 60]), you’ll get back the result of
10+20-30+40-50+60, or 50.

In [79]:
def plus_minus(seq):
    sign = 1
    sum = seq[0]
    for number in seq[1:]:
        sum = sum + number * sign
        sign = sign * -1
    return sum

plus_minus(range(10, 70, 10))

50

Write a function that partly emulates the built-in zip function (http://mng.bz/Jyzv), taking any number of iterables and returning a list of tuples. Each tuple will contain one element from each of the iterables passed to the function.
Thus, if I call myzip([10, 20,30], 'abc'), the result will be [(10, 'a'), (20,'b'), (30, 'c')]. You can return a list (not an iterator) and can assume that all of the iterables are of the same length

In [7]:
def my_zip(*iterables):
    result = []
    smallest = min([len(iterable) for iterable in iterables])
    for index in range(smallest):
        result.append(tuple([iterable[index] for iterable in iterables]))
    return result
    
my_zip([1, 2, 5], 'abcd', (7, 5))

[(1, 'a', 7), (2, 'b', 5)]

Write a function, sum_numeric, that takes any number of arguments. If the argument is or can be turned into an integer, then it should be added to the total. Arguments that can’t be handled as integers should be ignored. The
result is the sum of the numbers. Thus, sum_numeric(10, 20, 'a', '30', 'bcd') would return 60. Notice that even if the string 30 is an element in the list, it’s converted into an integer and added to the total.

In [13]:
def sum_numeric(*elements):
    total = 0
    for element in elements:
        try:
           if isinstance(element, int):
              total = total + element
           else:
              total = total + int(element)
        except:
           print('🔥🔥🔥🔥🔥🔥🔥')
           continue
    return total

sum_numeric(2, 6, 5, '8', 'kasozi', 0, '7g')

🔥🔥🔥🔥🔥🔥🔥
🔥🔥🔥🔥🔥🔥🔥


21

Write a function that takes a list of dicts and returns a single dict that combines all of the keys and values. If a key appears in more than one argument, the value should be a list containing all of the values from the arguments.

This challenge asks you to redefine the mysum function we defined in chapter 1, such that it can take any number of arguments. The arguments must all be of the same type and know how to respond to the + operator. (Thus, the function should work with numbers, strings, lists, and tuples, but not with sets and dicts.)

In [19]:
def my_sum(*args):
    result = args[0]
    for arg in args[1:]:
        result = result + arg
    return result
my_sum([1, 2, 8, 5], [6, 7])

[1, 2, 8, 5, 6, 7]

Write a function, mysum_bigger_than, that works the same as mysum, except that it takes a first argument that precedes *args. That argument indicates the threshold for including an argument in the sum. Thus, calling mysum_bigger
_than(10, 5, 20, 30, 6) would return 50—because 5 and 6 aren’t greater than 10. This function should similarly work with any type and assumes that all of the arguments are of the same type. Note that > and < work on many different types in Python, not just on numbers; with strings, lists, and tuples, it refers to their sort order.

Given a sequence of positive and negative numbers, sort them by absolute value.

In [22]:
def sort_by_abs(seq):
    return sorted(seq, key=abs)

[0, 2, -4, 8, -100]

Given a list of strings, sort them according to how many vowels they contain.

In [33]:
def count_vowels(word):
    return sum([1 for letter in word if is_vowel(letter)])

def count_as(word):
    return sum([1 for letter in word if letter == 'a'])

names = ['kasozi', 'vincent', 'vanessa', 'adrian']
sorted(names, key=count_as)

['vincent', 'kasozi', 'vanessa', 'adrian']

Given a list of lists, with each list containing zero or more numbers, sort by the sum of each inner list’s numbers.

In [37]:
sorted([[1, 5], [5, 6, 4], [9]], key=sum)

[[1, 5], [9], [5, 6, 4]]

Write a function, most_repeating_word, that takes a sequence of strings as input. The function should return the string that contains the greatest number of repeated letters. In other words
- For each word, find the letter that appears the most times.
- Find the word whose most-repeated letter appears more than any other.

Instead of finding the word with the greatest number of repeated letters, find the word with the greatest number of repeated vowels.

## Dictionaries

In this exercise, I want you to create a new constant dict, called MENU, representing
the possible items you can order at a restaurant. The keys will be strings, and the values will be prices (i.e., integers). You should then write a function, restaurant, that
asks the user to enter an order:
- If the user enters the name of a dish on the menu, the program prints the price
and the running total. It then asks the user again for their order.
- If the user enters the name of a dish not on the menu, the program scolds the
user (mildly). It then asks the user again for their order.
- If the user enters an empty string, the program stops prompting and prints the
total amount.

In [26]:
menu = {'tea': 10, 'pizza': 25, 'chawarma': 12, 'tacos': 20}  

def restaurant():
    total_price = 0
    while True:
       order = input('Enter your order: ')
       if order == '':
          print(f'The total is {total_price}')
          break
       elif order in menu:
           price = menu[order]
           total_price = total_price + price
           print(f'item: {order},price: {price}, total price: {total_price}')
       else:
           print('sorry, we don\'t have that')
          
restaurant()

The total is 0


Create a dict in which the keys are usernames and the values are passwords,both represented as strings. Create a tiny login system, in which the user must enter a username and password. If there is a match, then indicate that the user
has successfully logged in. If not, then refuse them entry. (Note: This is a nice little exercise, but please never store unencrypted passwords. It’s a major security risk.)

In [10]:
credentials = {'kasozi': 'function', 'mirembe': 'computer', 'bob': 'alice'}

def login():
    user_name = input('Enter your user name: ')
    password = input('Enter your password: ')
    if user_name in credentials:
        if password == credentials[user_name]:
            print('login successfull')
        else:
            print('Invalid password')
    else:
        print('user name doesn\'t exist')

login()

Invalid password


Define a dict whose keys are dates (represented by strings) from the most recent week and whose values are temperatures. Ask the user to enter a date, and display the temperature on that date, as well as the previous and subsequent dates, if available.

Define a dict whose keys are names of people in your family, and whose values are their birth dates, as represented by Python date objects (http://mng.bz/jggr). Ask the user to enter the name of someone in your family, and have the
program calculate how many days old that person is. 

In [18]:
from datetime import date

family_details = {
    'kasozi': date(1997, 9, 4),
    'vanessa': date(2001, 5, 25),
    'isaac': date(1642, 12, 25)
}

def get_days():
    name = input('Insert the name :')
    dob = family_details.get(name, date.today())
    today = date.today()
    days = (today - dob).days
    print(f'{name} is {days} days old')

get_days()

bob is 0 days old


Specifically, write a function, get_rainfall, that tracks rainfall in a number of cities. Users of your program will enter the name of a city; if the city name is blank, then the function prints a report (which I’ll describe) before exiting. If the city name isn’t blank, then the program should also ask the user how much rain has fallen in that city (typically measured in millimeters). After the user enters the quantity of rain, the program again asks them for a city name, rainfall amount, and so on—until the user presses Enter instead of typing the name of a city.When the user enters a blank city name, the program exits—but first, it reports
how much total rainfall there was in each city

In [23]:
def get_rainfall():
    rainfall = {}
    while True:
       city = input('Enter the city: ')
       if city == '':
          print_report(rainfall)
          break
       else:
          amount = int(input(f'Enter the amount of rainfall for {city}'))
          rainfall[city] = rainfall.get(city, 0) + amount


def print_report(rainfall):
   for city, amount in rainfall.items():
      print(f'{city} --------- {amount}')

get_rainfall()

Kampala --------- 50
Mbarara --------- 12


Instead of printing just the total rainfall for each city, print the total rainfall and the average rainfall for reported days. Thus, if you were to enter 30, 20, and 40 for Boston, you would see that the total was 90 and the average was 30.

In [16]:
def get_rainfall_avg():
    rainfall = {}
    while True:
        city = input('Enter the city: ')
        if city == '':
            print_report(rainfall)
            break
        else:
            amount = int(input('Enter the amount in millimeters: '))
            rainfall[city] = rainfall.get(city, []) + [amount]
            print(rainfall)

def print_report(rainfall):
    print('city: ------------- Total ---------------- Avg')
    for (key, value) in rainfall.items():
        print(f'{key} ------------- {sum(value)} ---------------- {sum(value)/len(value)}')

get_rainfall_avg()

city: ------------- Total ---------------- Avg


Read through a text file on disk. Use a dict to track how many words of each length are in the file—that is, how many three-letter words, four-letter words, five-letter words, and so on. Display your results.

In [37]:
def read_data():
    file = open('data/data.txt')
    for line in file.readlines():
       for word in line.strip().split():
        yield word

def count_words():
    word_count = {}
    for word in read_data():
        word_count[word[0]] = word_count.get(word[0], []) + [word]
    return word_count

count_words()

{'L': ['Lorem'],
 'i': ['ipsum',
  'id',
  'in',
  'iaculis',
  'in',
  'in',
  'interdum',
  'ipsum',
  'ipsum',
  'in',
  'in',
  'iaculis',
  'ipsum',
  'in',
  'id',
  'id',
  'id'],
 'd': ['dolor',
  'dictum',
  'diam',
  'diam',
  'dignissim',
  'dapibus',
  'dapibus',
  'dolor'],
 's': ['sit',
  'suscipit.',
  'suscipit',
  'sem',
  'sem',
  'scelerisque.',
  'semper.',
  'sit',
  'sed',
  'sapien',
  'sem.',
  'suscipit',
  'sed',
  'sed',
  'sapien',
  'sodales',
  'sapien.',
  'suscipit.',
  'sem'],
 'a': ['amet,',
  'adipiscing',
  'ac',
  'at',
  'augue',
  'a',
  'ac',
  'at',
  'aliquam',
  'a',
  'aliquam,',
  'at',
  'accumsan,',
  'ante',
  'ac',
  'a',
  'amet',
  'a',
  'ante',
  'a',
  'arcu',
  'aliquam',
  'a',
  'aliquet',
  'ac,',
  'aliquet',
  'at',
  'accumsan.',
  'ac',
  'ac',
  'aliquam',
  'aliquet.',
  'auctor,',
  'aliquet'],
 'c': ['consectetur',
  'consectetur,',
  'consectetur',
  'consectetur',
  'cubilia',
  'curae;',
  'consequat,',
  'commodo',
 

Open a log file from a Unix/Linux system—for example, one from the Apache server. For each response code (i.e., three-digit code indicating the HTTP request’s success or failure), store a list of IP addresses that generated that code.

In [48]:
def read_data():
    file = open('data/apache.txt')
    for line in file.readlines():
        yield line.strip()

def foo():
    D = {}
    for line in read_data():
        fields = line.split() #convert it to a list so that we can easily get the items
        ip_address = fields[0] # the first field is the ip address
        status_code = fields[8] # the eighth field is the http status code
        D[status_code] = D.get(status_code, []) + [ip_address]
    return D
foo()

{'200': ['192.168.1.101', '192.168.1.101'],
 '403': ['192.168.1.102'],
 '500': ['192.168.1.103'],
 '300': ['192.168.1.101'],
 '404': ['192.168.1.102']}

The dict.update method merges two dicts. Write a function that takes any number of dicts and returns a dict that reflects the combination of all of them. If the same key appears in more than one dict, then the most recently merged
dict’s value should appear in the output.

In [59]:
def merge(*dicts):
    result = {}
    for _dict in dicts:
        result.update(_dict)
    return result

merge({'name': 'kasozi', 'age': 26}, {'country': 'Algeria'})

{'name': 'kasozi', 'age': 26, 'country': 'Algeria'}

Write a function that takes any even number of arguments and returns a dict based on them. The even-indexed arguments become the dict keys, while the odd-numbered arguments become the dict values. Thus, calling the function with the arguments ('a', 1, 'b', 2) will result in the dict {'a':1, 'b':2} being returned.

In [57]:
def from_args(*args):
    if len(args) % 2 == 0:
        even_indexed = [i for i in args[::2]] # from 0 to end of list
        odd_indexed = [i for i in args[1::2]] # from 1 to end of list
        return dict(zip(even_indexed, odd_indexed))
    else:
        print('please pass in an even number of arguments')

from_args(1, ['a'], 5, 'f')

{(1,): ['a'], 5: 'f'}

Write a function , dict_partition, that takes one dict (d) and a function (f) as arguments. dict_partition will return two dicts, each containing key-value pairs from d. The decision regarding where to put each of the key-value pairs
will be made according to the output from f, which will be run on each key-value pair in d. If f returns True, then the key-value pair will be put in the first output dict. If f returns False, then the key-value pair will be put in the second output dict.

## Sets

Write a function, called how_many_different_numbers, that takes a single list of integers and returns the number of different integers it contains.

In [64]:
numbers = [1, 1, 6, 5, 5, 6, 3, 5, 4]

def count_unique(numbers):
    results = set(numbers)
    return list(results)

count_unique(numbers)

[1, 3, 4, 5, 6]

Read through a server (e.g., Apache or nginx) log file. What were the different IP addresses that tried to access your server?

In [9]:
def read_file(file_name):
    for line in open(file_name):
        yield line.split()[0]

def get_ip_addresses():
    ip_addresses = []
    for ip_address in read_file('data/apache.txt'):
        ip_addresses.append(ip_address)
    return list(set(ip_addresses)) # first convert it to a set to remove duplicates and then convert it to a list
        
get_ip_addresses()

['192.168.1.101', '192.168.1.102', '192.168.1.103']

Reading from that same server log, what response codes were returned to users? The 200 code represents “OK,” but there are also 403, 404, and 500 errors. (Regular expressions aren’t required here but will probably help.)

Use os.listdir (http://mng.bz/YreB) to get the names of files in the current directory. What file extensions (i.e., suffixes following the final . character) appear in that directory? It’ll probably be helpful to use os.path.splitext
(http://mng.bz/GV4v)

## Files

Write a python program that reads a text file and prints the last line in the text file

In [3]:
def print_lines(filename):
    with open(filename) as file:
        for line in file:
            print(line.strip())

print_lines('data/names.txt')

Namuyiga Phionah
Nabukenya Flavia
Namatovu Viola
kasozi Vincent
Mbidde Edward
Nakachwa Bridget
Bukenya Brian
Kimbugwe Vincent


write a python function that reads a file and returns the individual lines

In [4]:
def read_lines(filename):
    with open(filename, encoding='utf-8', errors='ignore') as file:
        for line in file:
            yield line.strip()

Write a python function that reads individual words from a text file

In [5]:
def read_words(filename):
    for line in read_lines(filename):
        for word in line.split(): # split the string using space into words
            yield word

Write a python function that reads a file and returns the letters in the file

In [6]:
def read_letters(filename):
    for word in read_words(filename):
        for letter in word:
            yield letter

Iterate over the lines of a text file. Find all of the words (i.e., non-whitespace surrounded by whitespace) that contain only integers, and sum them.

In [18]:
def sum_numbers(filename):
    total = 0
    for word in read_words(filename):
        if str.isdigit(word): # check to see if the string contains only numbers
            total = total + int(word)
    return total

sum_numbers('data/numbers.txt')

92

Read through a text file, line by line. Use a dict to keep track of how many times a given set of letters appears in the file. Print the resulting tabulation.

In [7]:
def frequencies(filename, predicate): # a predicate is a function that takes in args and returns a bool
    frequencies = {}
    for letter in read_letters(filename):
        if predicate(letter):
            frequencies[letter] = frequencies.get(letter, 0) + 1
    return frequencies

Read through a text file, line by line. Use a dict to keep track of how many times each vowel appears in the file. Print the resulting tabulation.

In [20]:
frequencies('data/apache.txt', is_vowel)

NameError: name 'is_vowel' is not defined

In this exercise, write a function, passwd_to_dict, that reads from a Unix-style “password file,” commonly stored as /etc/passwd, and returns a dict based on it. The function passwd_to_dict should return a dict based on /etc/passwd in
which the dict’s keys are usernames and the values are the users’ IDs.

In [ ]:
def passwd_to_dict(filename):
    details = {}
    for line in read_lines(filename):
        if line.startswith('#'):
            continue # when we see a comment, we just go to the next line
        else:
            fields = line.split(':')
            user_name = fields[0]
            user_id = fields[2]
            details[user_name] = user_id
    return details

passwd_to_dict('data/password.txt')

{'nobody': '-2',
 'root': '0',
 'daemon': '1',
 '_uucp': '4',
 '_taskgated': '13',
 '_networkd': '24',
 '_installassistant': '25',
 '_lp': '26',
 '_postfix': '27',
 '_scsd': '31',
 '_ces': '32',
 '_mcxalr': '54',
 '_appleevents': '55',
 '_geod': '56',
 '_serialnumberd': '58',
 '_devdocs': '59',
 '_sandbox': '60',
 '_mdnsresponder': '65',
 '_ard': '67',
 '_www': '70',
 '_eppc': '71',
 '_cvs': '72',
 '_svn': '73',
 '_mysql': '74',
 '_sshd': '75',
 '_qtss': '76',
 '_cyrus': '77',
 '_mailman': '78',
 '_appserver': '79',
 '_clamav': '82',
 '_amavisd': '83',
 '_jabber': '84',
 '_appowner': '87',
 '_windowserver': '88',
 '_spotlight': '89',
 '_tokend': '91',
 '_securityagent': '92',
 '_calendar': '93',
 '_teamsserver': '94',
 '_update_sharing': '95',
 '_installer': '96',
 '_atsserver': '97',
 '_ftp': '98',
 '_unknown': '99',
 '_softwareupdate': '200',
 '_coreaudiod': '202',
 '_screensaver': '203',
 '_locationd': '205',
 '_trustevaluationagent': '208',
 '_timezone': '210',
 '_lda': '211',
 '_c

The challenge for this exercise is to write a wordcount function that mimics the wc
Unix command. The function will take a filename as input and will print four lines
of output:
1. Number of characters (including whitespace)
2. Number of words (separated by whitespace)
3. Number of lines
4. Number of unique words (case sensitive, so “NO” is different from “no”)

In [ ]:
def word_count(filename):
    return {
        'characters': len(list(read_letter(filename))),
        'words': len(list(read_words(filename))),
        'lines': len(list(read_lines(filename))),
        'unique': len(set(read_words(filename)))
    }

word_count('data/server.log')

{'characters': 33367, 'words': 3195, 'lines': 206, 'unique': 593}

Ask the user to enter the name of a text file and then (on one line, separated by spaces) words whose frequencies should be counted in that file. Count how many times those words appear in a dict, using the user-entered words as the
keys and the counts as the values.

In [ ]:
def word_frequencies(filename=r'data\text.txt'):
    results = {}
    words = input('Enter the words: ')
    words = words.strip().split()

    for word in read_words(filename.strip()):
        if word in words:
            results[word] = results.get(word, 0) + 1
    return results
word_frequencies()

{'and': 21, 'a': 7, 'is': 3}

Create a dict in which the keys are the names of files on your system and the values are the sizes of those files. To calculate the size, you can use os.stat (http://mng.bz/dyyo).

In [ ]:
import os
def file_sizes(folder_name):
    details = {}
    for filename in os.listdir(folder_name):
        details[filename] = os.stat(f'{folder_name}\{filename}').st_size
    return details

file_sizes('files')

{'file_1.txt': 5098, 'file_2.txt': 619, 'file_3.txt': 3343}

Given a directory, read through each file and count the frequency of each letter. (Force letters to be lowercase, and ignore nonletter characters.) Use a dict to keep track of the letter frequencies. What are the five most common letters
across all of these files?

In [13]:
import os
# make the letters lowercase str.lower
# ignore non letter characters str.isalpha

# create the function that works on one file.
def letter_frequencies(filename):
    frequencies = {}
    for letter in read_letters(filename):
        if letter.isalpha():
            s = letter.lower() # make the letter small case
            frequencies[s] = frequencies.get(s, 0) + 1
    return frequencies

def process_files(foldername):
    results = {}
    for filename in os.listdir(foldername):
        full_filename = f'{foldername}\{filename}'
        results[full_filename] = letter_frequencies(full_filename)
    return results


process_files('data')

In this exercise, write two functions. find_longest_word takes a filename as an argument and returns the longest word found in the file. The second function, find_all_longest_words, takes a directory name and returns a dict in which the keys are filenames and the values are the longest words from each file

In [26]:
def find_longest_word(file_name):
    words = []
    for word in read_words(file_name):
        if word.isalpha():
           words.append(word.lower())
    return max(words, key=len)

def find_all_longest_words(folder_name):
    results = {}
    for filename in os.listdir(folder_name):
        full_file_name = f'{folder_name}\{filename}'
        results[full_file_name] = find_longest_word(full_file_name)
    return results


find_all_longest_words('data')

{'data\\apache.txt': 'like',
 'data\\data.txt': 'consectetur',
 'data\\names.txt': 'nabukenya',
 'data\\numbers.txt': 'fkjfsdkfjslkfjsklj',
 'data\\password.txt': 'configuration',
 'data\\server.log': 'windows',
 'data\\text.txt': 'cardiovascular'}

Use the hashlib module in the Python standard library, and the md5 function within it, to calculate the MD5 hash for the contents of every file in a user specified directory. Then print all of the filenames and their MD5 hashes.

In [34]:
from hashlib import md5

def hash_file_contents(folder_name):
    for filename in os.listdir(folder_name):
        full_file_name = f'{folder_name}\{filename}'
        print(full_file_name, md5(open(full_file_name, 'rb').read()).digest())
hash_file_contents('data')

Open an HTTP server’s log file. (If you lack one, then you can read one from me at http://mng.bz/vxxM.) Summarize how many requests resulted in numeric response codes—202, 304, and so on.

In [45]:
from toolz import frequencies
def process_file(file_name):
    codes = []
    for line in read_lines(file_name):
        status_code = line.split()[8]
        codes.append(status_code)
    return frequencies(codes)

process_file('data/server.log')

{'200': 193, '302': 6, '304': 5, '404': 2}

For this exercise, create a function, passwd_to_csv, that takes two filenames as
arguments: the first is a passwd-style file to read from, and the second is the name of a
file in which to write the output. <br>

The new file’s contents are the username (index 0) and the user ID (index 2).
Note that a record may contain a comment, in which case it will not have anything at
index 2; you should take that into consideration when writing the file. The output file
should use TAB characters to separate the elements

In [69]:
import csv
def passwd_to_csv(input_file_name, output_file_name):
    with open(input_file_name) as input, open(output_file_name, 'w', newline='') as output:
        reader = csv.reader(input, delimiter=':')
        writer = csv.writer(output)

        for record in reader:
            if len(record) > 1:
              writer.writerow([record[0], record[2]])

passwd_to_csv('data/password.txt', 'vanessa.csv')

Extend this exercise by asking the user to enter a space-separated list of integers, indicating which fields should be written to the output CSV file. Also ask the user which character should be used as a delimiter in the output file. Then read from /etc/passwd, writing the user’s chosen fields, separated by the user’s chosen delimiter.

In [10]:
import csv
def process_file(in_filename, out_filename):
    choice = input('Insert the fileds that you want to extract')
    indices = choice.strip().split()
    with open(in_filename) as _input, open(out_filename, 'w', newline='') as _output:
        reader = csv.reader(_input, delimiter=':')
        writer = csv.writer(_output)
        for record in reader:
            if len(record) > 1:
                fields = [record[int(index)] for index in indices]
                writer.writerow(fields)
process_file('data/password.txt', 'vanessa.csv')

Write a function that writes a dict to a CSV file. Each line in the CSV file should contain three fields: (1) the key, which we’ll assume to be a string, (2) the value, and (3) the type of the value (e.g., str or int).

In [13]:
def dict_to_csv(_dict, filename):
    with open(filename, 'w', newline='') as output:
        writer = csv.writer(output)
        for (key, value) in _dict.items():
            fields = [key, value, str(type(value))]
            writer.writerow(fields)

person = {
    'first_name': 'kasozi',
    'second_name': 'vincent',
    'age': 26,
    'weight': 42.6,
    'is_girl': False
}

dict_to_csv(person, 'person.csv')

Create a CSV file, in which each line contains 10 random integers between 10 and 100. Now read the file back, and print the sum and mean of the numbers on each line. 

In [17]:
from random import randint

def create_csv(filename):
    with open(filename, 'w', newline='') as csv_file:
        writer = csv.writer(csv_file) 
        for row in range(1,11):
            writer.writerow([randint(10,100) for _ in range(1,11)])

def read_csv(filename):
    with open(filename) as csv_file:
        reader = csv.reader(csv_file)
        for record in reader:
            record = [int(field) for field in record]
            print(f'sum: {sum(record)} avg: {sum(record)/len(record)}')
        
read_csv('numbers.csv')

sum: 381 avg: 38.1
sum: 535 avg: 53.5
sum: 436 avg: 43.6
sum: 574 avg: 57.4
sum: 663 avg: 66.3
sum: 414 avg: 41.4
sum: 543 avg: 54.3
sum: 577 avg: 57.7
sum: 593 avg: 59.3
sum: 615 avg: 61.5


# Functions

Write a function, myxml, that allows you to create simple XML output. The output from the function will always be a string.

In [41]:
def dict_to_str(_dict):
    attribs = []
    for (key, value) in _dict.items():
        attribs.append(f'{key}="{value}"')
    return ' '.join(attribs)

def my_xml(tag_name, content='', **kwargs): 
    print(kwargs)
    attributes = dict_to_str(kwargs['kwargs'])
    return f'<{tag_name} {attributes}>{content}</{tag_name}>'
my_xml(tag_name='person', content='kasozi vincet', kwargs={'title': 'mr', 'age':26})

{'kwargs': {'title': 'mr', 'age': 26}}


'<person title="mr" age="26">kasozi vincet</person>'

In [33]:
d = {
    'name': 'kasozi',
    'age': 4
}

for (key, value) in d.items():
    print(f'{key}="{value}"')

name="kasozi"
age="4"


Write a copyfile function that takes one mandatory argument—the name of an input file—and any number of additional arguments: the names of files to which the input should be copied. Calling copyfile('myfile.txt', 'copy1.txt', 'copy2.txt', 'copy3.txt') will create three copies of myfile.txt: one each in copy1.txt, copy2.txt, and copy3.txt.

Write a “factorial” function that takes any number of numeric arguments and returns the result of multiplying them all by one another.

Write an anyjoin function that works similarly to str.join, except that the first argument is a sequence of any types (not just of strings), and the second argument is the “glue” that we put between elements, defaulting to " " (a space). So anyjoin([1,2,3]) will return 1 2 3, and anyjoin('abc', pass:'**') will return pass:a**b**c.